onnx/helper.py

```diff
@@ -690,8 +690,7 @@ def pack_float32_to_float4e2m1(array: np.ndarray | Sequence) -> np.ndarray:
     if is_odd_volume:
         array_flat = np.append(array_flat, np.array([0]))
 
-    func = np.vectorize(subbyte.float32x2_to_float4e2m1x2)
-    arr = func(array_flat[0::2], array_flat[1::2])
+    arr = subbyte.float32x2_to_float4e2m1x2(array_flat[0::2], array_flat[1::2])
```

Comment:
which ones should be in numpy_helper and which ones should be in subbyte? Any thoughts?

Reply:
The boundary is a bit blurry. The functions in subbyte are also numpy helper functions - only that they do manipulations at the subbyte level. That said, some functions in numpy_helper also deal with packing/unpacking.

I guess we could do some reorganization for the subbyte functions in a subsequent MR, as there are other changes proposed as well in https://github.com/onnx/onnx/pull/6124.

Reply:
And `float32x2` is just indicating that there are two input float32's. I think it'd be helpful to keep that info in the name.

Reply:
I now think subtype should be an internal module, as it is not ONNX responsibility to provide type conversion utilities for numpy. (numpy_helper provides utilities for creating tensor protos from numpy arrays).

We should consider changing subtype to internal (in a separate PR). Users can copy the functions out if they rely on them as dependencies.

I suggest we move all subtype related, np -> np functions in subtype, and define them privately (append `_`). In the next PR we will make subtype itself private and make the functions public within the onnx package so their visibility is proper.

Reply:
> And `float32x2` is just indicating that there are two input float32's. I think it'd be helpful to keep that info in the name.

That's reasonable. Although I would suggest that we do

1. cast float32 to float4 (unpacked, as uint8)
2. pack uint8 to uint4x2 (aka float4x2)

so that the logic is very clear, and we can avoid this `float32x2_to_float4e2m1x2` function.

Reply:
The current structure parallels that of int4/uint4, ie there are:

- `float32_to_float4e2m1_unpacked` and `float32_to_4bit_unpacked` which perform casting to fp4/int4 (unpacked as uint8)
- `float32x2_to_float4e2m1x2` and `float32x2_to_int4x2`  which call the above functions as do the uint8->uint4x2 packing (which is a one-liner).
- `pack_float32_to_float4e2m1` and `pack_float32_to_4bit` which does necessary padding and casting and call the above functions.

I think if we decide to refactor the logic we can do make the changes in a follow-up PR for both fp4 and int4/uint4 to keep the structure parallel.

Reply:
Sounds good. We should refactor right after this. https://github.com/onnx/onnx/issues/6296